// tslint-etc/dist/support/tslint-deprecation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var tsutils = require("tsutils");
var ts = require("typescript");
function isDeclaration(identifier) {
    var parent = identifier.parent;
    switch (parent.kind) {
        case ts.SyntaxKind.ClassDeclaration:
        case ts.SyntaxKind.ClassExpression:
        case ts.SyntaxKind.InterfaceDeclaration:
        case ts.SyntaxKind.TypeParameter:
        case ts.SyntaxKind.FunctionExpression:
        case ts.SyntaxKind.FunctionDeclaration:
        case ts.SyntaxKind.LabeledStatement:
        case ts.SyntaxKind.JsxAttribute:
        case ts.SyntaxKind.MethodDeclaration:
        case ts.SyntaxKind.MethodSignature:
        case ts.SyntaxKind.PropertySignature:
        case ts.SyntaxKind.TypeAliasDeclaration:
        case ts.SyntaxKind.GetAccessor:
        case ts.SyntaxKind.SetAccessor:
        case ts.SyntaxKind.EnumDeclaration:
        case ts.SyntaxKind.ModuleDeclaration:
            return true;
        case ts.SyntaxKind.VariableDeclaration:
        case ts.SyntaxKind.Parameter:
        case ts.SyntaxKind.PropertyDeclaration:
        case ts.SyntaxKind.EnumMember:
        case ts.SyntaxKind.ImportEqualsDeclaration:
            return parent.name === identifier;
        case ts.SyntaxKind.PropertyAssignment:
            return (parent.name === identifier &&
                !tsutils.isReassignmentTarget(identifier.parent
                    .parent));
        case ts.SyntaxKind.BindingElement:
            return (parent.name === identifier &&
                parent.propertyName !== undefined);
        default:
            return false;
    }
}
exports.isDeclaration = isDeclaration;
function getCallExpresion(node) {
    var parent = node.parent;
    if (tsutils.isPropertyAccessExpression(parent) && parent.name === node) {
        node = parent;
        parent = node.parent;
    }
    return tsutils.isTaggedTemplateExpression(parent) ||
        ((tsutils.isCallExpression(parent) || tsutils.isNewExpression(parent)) &&
            parent.expression === node)
        ? parent
        : undefined;
}
function getDeprecation(node, tc) {
    var callExpression = getCallExpresion(node);
    if (callExpression !== undefined) {
        var result = getSignatureDeprecation(tc.getResolvedSignature(callExpression));
        if (result !== undefined) {
            return result;
        }
    }
    var symbol;
    var parent = node.parent;
    if (parent.kind === ts.SyntaxKind.BindingElement) {
        symbol = tc.getTypeAtLocation(parent.parent).getProperty(node.text);
    }
    else if ((tsutils.isPropertyAssignment(parent) && parent.name === node) ||
        (tsutils.isShorthandPropertyAssignment(parent) &&
            parent.name === node &&
            tsutils.isReassignmentTarget(node))) {
        symbol = tc.getPropertySymbolOfDestructuringAssignment(node);
    }
    else {
        symbol = tc.getSymbolAtLocation(node);
    }
    if (symbol !== undefined &&
        tsutils.isSymbolFlagSet(symbol, ts.SymbolFlags.Alias)) {
        symbol = tc.getAliasedSymbol(symbol);
    }
    if (symbol === undefined ||
        (callExpression !== undefined && isFunctionOrMethod(symbol.declarations))) {
        return undefined;
    }
    return getSymbolDeprecation(symbol);
}
exports.getDeprecation = getDeprecation;
function findDeprecationTag(tags) {
    var e_1, _a;
    try {
        for (var tags_1 = tslib_1.__values(tags), tags_1_1 = tags_1.next(); !tags_1_1.done; tags_1_1 = tags_1.next()) {
            var tag = tags_1_1.value;
            if (tag.name === "deprecated") {
                return tag.text === undefined ? "" : tag.text;
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (tags_1_1 && !tags_1_1.done && (_a = tags_1.return)) _a.call(tags_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return undefined;
}
function getSymbolDeprecation(symbol) {
    if (symbol.getJsDocTags !== undefined) {
        return findDeprecationTag(symbol.getJsDocTags());
    }
    return getDeprecationFromDeclarations(symbol.declarations);
}
function getSignatureDeprecation(signature) {
    if (signature === undefined) {
        return undefined;
    }
    if (signature.getJsDocTags !== undefined) {
        return findDeprecationTag(signature.getJsDocTags());
    }
    return signature.declaration === undefined
        ? undefined
        : getDeprecationFromDeclaration(signature.declaration);
}
function getDeprecationFromDeclarations(declarations) {
    var e_2, _a;
    if (declarations === undefined) {
        return undefined;
    }
    var declaration;
    try {
        for (var declarations_1 = tslib_1.__values(declarations), declarations_1_1 = declarations_1.next(); !declarations_1_1.done; declarations_1_1 = declarations_1.next()) {
            declaration = declarations_1_1.value;
            if (tsutils.isBindingElement(declaration)) {
                declaration = tsutils.getDeclarationOfBindingElement(declaration);
            }
            if (tsutils.isVariableDeclaration(declaration)) {
                declaration = declaration.parent;
            }
            if (tsutils.isVariableDeclarationList(declaration)) {
                declaration = declaration.parent;
            }
            var result = getDeprecationFromDeclaration(declaration);
            if (result !== undefined) {
                return result;
            }
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (declarations_1_1 && !declarations_1_1.done && (_a = declarations_1.return)) _a.call(declarations_1);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return undefined;
}
function getDeprecationFromDeclaration(declaration) {
    var e_3, _a, e_4, _b;
    try {
        for (var _c = tslib_1.__values(tsutils.getJsDoc(declaration)), _d = _c.next(); !_d.done; _d = _c.next()) {
            var comment = _d.value;
            if (comment.tags === undefined) {
                continue;
            }
            try {
                for (var _e = (e_4 = void 0, tslib_1.__values(comment.tags)), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var tag = _f.value;
                    if (tag.tagName.text === "deprecated") {
                        return tag.comment === undefined ? "" : tag.comment;
                    }
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
        }
        finally { if (e_3) throw e_3.error; }
    }
    return undefined;
}
function isFunctionOrMethod(declarations) {
    if (declarations === undefined || declarations.length === 0) {
        return false;
    }
    switch (declarations[0].kind) {
        case ts.SyntaxKind.MethodDeclaration:
        case ts.SyntaxKind.FunctionDeclaration:
        case ts.SyntaxKind.FunctionExpression:
        case ts.SyntaxKind.MethodSignature:
            return true;
        default:
            return false;
    }
}
